### DISCUSSION TITLE
Templates for Single Product Types

### DISCUSSION THREAD

**Discussion by nefeline** [woocommerce/woocommerce#44956] [WooCommerce, Blocks, and Site Editing]
As part of the [Blockifying Single Product Template](https://github.com/woocommerce/woocommerce-blocks/issues/8054) project, @albarin, @kmanijak, and I kicked off an investigation on different alternatives for implementing single product templates for each one of the product types.

We started this work by identifying what product types are in the scope of WordPress and the possible paths considering the extendability of the native template hierarchy. In a nutshell, product types are terms of the `product_type` taxonomy:

```sql
SELECT slug
FROM wp_terms
         INNER JOIN wp_term_taxonomy wtt ON wp_terms.term_id = wtt.term_id
WHERE wtt.taxonomy = 'product_type'
```
<img width="173" alt="Screenshot 2023-02-14 at 10 36 51" src="https://user-images.githubusercontent.com/15730971/219980480-26b00738-c93c-4d6b-a3b8-6232e901dba7.png">

When we consider the [default template hierarchy for a single post](https://developer.wordpress.org/themes/basics/template-hierarchy/#single-post), this is the path followed by WordPress:

* single-{post-type}-{slug}.php – First, WordPress looks for a template for the specific post. For example, if post type is product and the post slug is dmc-12, WordPress would look for single-product-dmc-12.php.
* single-{post-type}.php – If the post type is product, WordPress would look for single-product.php.
* single.php – WordPress then falls back to single.php.
* singular.php – Then it falls back to singular.php.
* index.php – Finally, as mentioned above, WordPress ultimately falls back to index.php.

In other words: WordPress, out-of-the-box, doesn’t provide a template for a single post type that belongs to a specific taxonomy and term, so we had to come up with alternatives to circumvent this, by relying on the native hooks, such as the `'single_template_hierarchy'` filter and alternative solutions.

In the following section, we describe each of the short and long-term solutions evaluated, as well as their pros and cons.

# Short-term alternatives

## [Register templates for each product type](https://github.com/woocommerce/woocommerce-blocks/pull/8376)

The first step of this approach is modifying the “single-product” template hierarchy. Using the [single_template_hierarchy](https://developer.wordpress.org/reference/hooks/type_template_hierarchy/) hook we [introduce a new template](https://github.com/woocommerce/woocommerce-blocks/pull/8376/files#diff-8ae3c446b897aa3d87c4814f390b6809b5af8cbf0e29180adb945afe6dfd6445R85) in the second position of the array corresponding to the product type:

Before:

```js
array(3) {
 [0]=> string(25) "single-product-beanie.php" 
 [1]=> string(18) "single-product.php" 
 [2]=> string(10) "single.php" 
}
```

After:

```js
array(3) {
 [0]=> string(25) "single-product-beanie.php" 
 [1]=> string(18) "single-product-{TYPE}"
 [2]=> string(18) "single-product.php" 
 [3]=> string(10) "single.php" 
}
```
That way, if there’s a template for a product type (e.g. “single-product-simple”) it will take precedence over the more general “single-product” template.

The second step is using the [get_block_file_template](https://developer.wordpress.org/reference/functions/get_block_file_template/) and [get_block_templates](https://developer.wordpress.org/reference/functions/get_block_templates/) hook to ensure we register a template for each of the available product types in the Site Editor.
If no template exists (in the theme or the database) for a product type, we’ll use the “single-product” template as a fallback.

As a result, the Site Editor will look as follows:

<img width="638" alt="Screenshot 2023-02-20 at 00 14 36" src="https://user-images.githubusercontent.com/15730971/219981165-ea038cb9-2079-4e54-9218-da20ded9e49b.png">

|Pros |Cons|
|-----|--------|
| Simple solution that uses available hooks. | The UI may not be the best and could end up showing a very long list of templates if there are lots of product types. |

## [Register product types as post types so they display as templates](https://github.com/woocommerce/woocommerce-blocks/pull/8379)

One of the entities displayed in Gutenberg’s Template menu is Post Types. These are recognized as `postTypesMenuItems`. All post types that are publicly_queryable will be displayed in the Add New Template submenu.

| ![219981615-5270977e-3a44-46ae-905b-ed6975c96c5b.png](https://user-images.githubusercontent.com/15730971/219981615-5270977e-3a44-46ae-905b-ed6975c96c5b.png) | 
|:--:| 
| _Post types used to present templates to user: - blue - default ones: Post. Green - additional post types that apear in submenu (the one after clicking “+” button): Single item: Post and Single item: Product. Red - newly attached post types._ |

Given the above, each product type can have a separate template if we register it as a post type. This approach is based on the [register_post_type](https://developer.wordpress.org/reference/functions/register_post_type/) function that allows doing that. When initializing block templates, we can call the function for each of the available product types. The function allows providing multiple options, like a custom label or description.

Important: for the post type to appear in the Add New Template submenu, the `publicly_queryable` option must be set to true.

<img width="242" alt="Screenshot 2023-02-20 at 00 34 03" src="https://user-images.githubusercontent.com/15730971/219982140-830d87fb-6e1a-41e4-b81a-eba47b61ffad.png">

_End result on the UI. It reflects the data presented in the JS object in the previous figure._

|Pros |Cons|
|-----|--------|
| It works! | Bad UX: As you can see in the video attached to the PR, “custom” post types appear in the menu with the couple of seconds delay. Additionally, the order of these varies between sessions: one time Category (product_cat) may appear above, sometimes below product types. |
|  |Hacky solution: This solution abuses a domain, by making product types something they are not. It may have some other consequences, that we didn’t discover yet. |
| | Requires further work: As default, they use a post template we didn’t manage to force them to use a single product template instead |

# Long-term solutions

## [Use the “extra_template_types” hook](https://github.com/woocommerce/woocommerce-blocks/pull/8358)

This approach uses the new “extra_templates_hook” that we want to introduce on Gutenberg on [this PR](https://github.com/WordPress/gutenberg/pull/46362) to add new items on the Site Editor “Add new” menu.
This new hook allows us to create and define new templates; in this case, [we would add one](https://github.com/woocommerce/woocommerce-blocks/pull/8358/files#diff-8ae3c446b897aa3d87c4814f390b6809b5af8cbf0e29180adb945afe6dfd6445R101) for each of the existing product types.

We would also need to [update the single product template hierarchy](https://github.com/woocommerce/woocommerce-blocks/pull/8358/files#diff-8ae3c446b897aa3d87c4814f390b6809b5af8cbf0e29180adb945afe6dfd6445R86) exactly in the same way we did on [this approach](https://github.com/woocommerce/woocommerce-blocks/pull/8376).

<img width="250" alt="Screenshot 2023-02-20 at 00 43 22" src="https://user-images.githubusercontent.com/15730971/219982621-55ea2d23-db8b-4785-a023-f88e4573123c.png">

Pros | Cons
-- | --
Very simple solution in terms of code. Cleaner UI experience. | The PR on Gutenberg is a blocker, and we don’t know how long it will take to pass code review. The UI is still not great since it shows one item per product type in the menu, which could grow a lot.|

## [Have the templates for each taxonomy displayed under the Single Item in the Site Editor.](https://github.com/WordPress/gutenberg/issues/48264)

This is an approach that allows the creation of a new template for a single post with a specific taxonomy and term:

<img width="528" alt="Screenshot 2023-02-20 at 00 52 43" src="https://user-images.githubusercontent.com/15730971/219983064-11054e7f-5db7-455f-9bff-f0c7e52312fc.png">

This implementation path is suggested as we believe the use case explored here is not exclusive to WooCommerce’s custom post types, but applicable to virtually any single post the site owner wishes to customize based on its assigned taxonomies and terms. These customization options are all possible nowadays by relying on hooks provided by WordPress on the PHP side, but as we migrate to an FSE world, expanding the template types that can be added for single posts could be a good step for the Site Editor in the near future.

There’s already an open issue to dynamically generate query variations for registered post types, based on the taxonomies attached to them:

https://github.com/WordPress/gutenberg/issues/40941

It would be great to rely on this implementation in the long run for generating the templates, although to be able to do so, we would also need the product type taxonomy to become public on Woo (which we already started to discuss on per0F9-ci-p2#comment-233 ).

|Pros |Cons|
|-----|--------|
| Better UX and alignment with what is planned for the core of WordPress. | The product type taxonomy needs to be public. |

# The Extensibility of templates in the site editor 

One of the most important aspects we noticed while working on the site editor and attempting to introduce our custom templates is that this area of Gutenberg still doesn’t have much to offer when it comes to customizations. The idea of making this area more extensible is not new:

https://github.com/WordPress/gutenberg/issues/41401

As Matias mentions:

> There are many plugins that come with their own set of templates (plugins for commerce, newsletters, education, etc) that would need to extend the list of templates available when creating new ones. So this list needs to be easy to filter and extend. From a design perspective, we also need to group them well both for creation flows (in the dropdown) and for management flows.

With that said, we believe https://github.com/WordPress/gutenberg/pull/46362 is well aligned with this goal and unlocks customization paths that are beneficial not only for us but for the community and plugin developers in general.

Additionally, it would be interesting to know when implementations such as the one proposed on https://github.com/WordPress/gutenberg/issues/40941 are planned to be developed and released by the Gutenberg team, as it can also serve our purposes here in a different way.

# Conclusion

Having custom templates for each one of the product types in the editor is possible. In the short term, there are two workarounds available for us to explore, but with the UX and development limitations highlighted in the previous sections.

In the long term, ideally, we would like to rely on Gutenberg and extend its native UI so that custom templates can be added in different areas, both in the “add new template” dropdown menu and single posts with specific taxonomies and terms.

**Comment by gigitux** [comment]
Thanks for taking care of this exploration and writing this excellent summary, folks! 🙇  

### Short-term solution

I think that we should follow @albarin's approach. As you said, the second solution looks like a "hacky approach". We are creating some types even if the product types are taxonomy.

> Bad UX: As you can see in the video attached to the PR, “custom” post types appear in the menu with the couple of seconds delay. Additionally, the order of these varies between sessions: one time Category (product_cat) may appear above, sometimes below product types.

This depends on Gutenberg. Also, the issue is visible on other templates too.

### Long-term solution

Sorry, but I don't understand the relation between this issue and https://github.com/WordPress/gutenberg/issues/40941. It seems that this latter belongs to the Query Loop block 🤔 

Furthermore, I think that @dinhtungdu's [comment](https://github.com/WordPress/gutenberg/pull/46362#issuecomment-1413619994) highlights very well the issue:

> Now we want to create templates for product types, we can't do that because Gutenberg only supports default entities that are used in the intended way (post type for a single item, taxonomy term for a list of items), we want to create the template for each product type which displays a single product but the entity powering it is a taxonomy, which for list of items.

Does creating a dedicated issue on the Gutenberg repo and seeing the feedback make sense?

**Comment by nefeline** [comment]
Thanks for bringing up these points for discussion @gigitux !

> I think that we should follow @albarin's approach. As you said, the second solution looks like a "hacky approach". We are creating some types even if the product types are taxonomy.

Yep, agreed (I also brought this up during the exploration phase p1675434706418459/1675247944.988549-slack-C02FL3X7KR6 ).

> Sorry, but I don't understand the relation between this issue and https://github.com/WordPress/gutenberg/issues/40941. It seems that this latter belongs to the Query Loop block 🤔 

Sorry, I should have been more clear here: this issue was brought up motivated by James' comment on https://github.com/WordPress/gutenberg/pull/46362#issuecomment-1344454425 as he mentions the approach where, if a custom post type or taxonomy is public then it automatically becomes possible to add templates for them & the same functionality could theoretically drive https://github.com/WordPress/gutenberg/issues/40941. I gather this issue doesn't fully cover what we are proposing as the long-term solution here, so a new one was created: https://github.com/WordPress/gutenberg/issues/48264

> Furthermore, I think that @dinhtungdu's https://github.com/WordPress/gutenberg/pull/46362#issuecomment-1413619994 highlights very well the issue

Yep, and this comment doesn't differ from our conclusion and what was described as the issue in the discussion above.

> Does creating a dedicated issue on the Gutenberg repo and seeing the feedback make sense?

Good point! It does: as mentioned previously, opened https://github.com/WordPress/gutenberg/issues/48264

**Reply by gigitux** [reply]
Thanks for the explanation! Great work, folks! 🙇

**Comment by dinhtungdu** [comment]
Thank you so much for this excellent write-up!

I totally agree with the long-term direction here. About the short term, I want to share my 2 cents:

I want to propose another option: that we do ... nothing:
- Currently, WooCommerce doesn't have templates for product types, so that isn't a functionality people are using.
- AFAIK, the most important difference between product types is the add-to-cart form, which is handled by block.
- By not introducing any temporary workaround, we save our limited resources and keep our codebase clean. We also won't need to migrate from the short-term to the long-term solution.
- Don't get wet before it rains (thanks again @sunyatasattva). Is there any way we can study/research if merchants/themes really need this feature? How many feature requests for this feature have we received?
- This feature isn't a blocker for any other. So I think we can, at least, leave it for later and focus on higher-priority features for now.

**Reply by gigitux** [reply]
The technical implementation of the templates for product type is one of the goals of this cycle. I think that it makes sense to work in a short-term solution since that:
- [we already have the implementation ready](https://github.com/woocommerce/woocommerce-blocks/pull/8376)
- it could be a nice feature that could convince folks to use a block theme.

cc @nerrad

**Reply by nerrad** [reply]
For the past cycle the primary outcome I was hoping for is a better understanding of:

* What would it take to implement Product Type specific templates?
* Are there any technical hurdles requiring work upstream?

Thanks to the explorations of the team, it's clear that while Product Type specific templates are doable, for a great user experience there is definitely some work needed upstream. 

As [Tung's suggestion](https://github.com/woocommerce/woocommerce-blocks/discussions/8475#discussioncomment-5060691) points out, not doing anything here is certainly a viable option. However, the thinking behind product-type-specific templates is that it improves the user experience around the variant layout and behaviour options between product types (as opposed to trying to manage that all via the single product template). It'd also be a great path for extensions that add _new_ product types to simply offer their own template for the single product view.

With that said, I don't think we have to _prioritize_ this work until we have some of the necessary upstream issues addressed - it's sufficient for now to know that it's technically feasible to support product types (with the appropriate UX improvements).

**Comment by Marc-pi** [comment]
Coming here thanks to @gigitux from https://github.com/woocommerce/woocommerce-blocks/issues/10363
What's the plan now for Grouped Products ?
Actually, the child elements are displayed through the Add to Cart With Options block, but there are no options (yet), for instance to be able to : 
- adjust the way child elements are displayed within it
- adjust child name font size
- add an Add to cart per child lines

Is there a plan for the short term to finalisze the `Add to Cart With Options` block, so we gain modularity with ... `options` and nested blocks ?
What's the target finally : a dedicated template for each product type ?

**Reply by Aljullu** [reply]
> Is there a plan for the short term to finalisze the `Add to Cart With Options` block, so we gain modularity with ... `options` and nested blocks ?

Hey there, @Marc-pi! We have started early explorations on blockifying the _Add to Cart With Options_ block. I wouldn't say it's something that will be solved in the short term, but if you are interested, you can have a look at this discussion: https://github.com/woocommerce/woocommerce/discussions/49799.

**Reply by Marc-pi** [reply]
thanks, i've just read your exploration, good to know you are working on it
